### DISCUSSION TITLE
Workarounds to mask port usage due to active queue?

### DISCUSSION THREAD

**Discussion by wlandau** [r-lib/mirai#31] [Q&A] [answered]
https://github.com/shikokuchuo/mirai/discussions/25#discussioncomment-5168174 and https://github.com/shikokuchuo/mirai/issues/27#issuecomment-1450468776 make me really optimistic and excited about using `mirai` for `crew`. The one remaining obstacle in my case is still the issue of large port ranges required by the client for the active queue (c.f. #26, 24, https://github.com/shikokuchuo/mirai/discussions/22#discussioncomment-5124391). Knowing what I know about the [`targets`](https://docs.ropensci.org/targets) user base, many users of `crew` will be sharing the client machine and IP address with dozens of colleagues, and so occupying hundreds of ports at a time may consume a sizable fraction of shared resources.

With an active queue, the servers dial directly into the active queue daemon instead of the actual client, right? What if the active queue were to run inside a VM or Docker/Singularity container? Wouldn't that create a whole new IP address and port range without risk of conflicting with other users? Is there a simpler and more platform-independent and idiomatic way to give the active queue daemon complete freedom and flexibility with ports? NAT and port forwarding seem like similar concepts but perhaps not helpful here.

**Comment by wlandau** [comment]
On the bright side, the real endgame is the cloud. If the client and active queue run in their own instance in a VPC, port usage will not be a problem.

**Reply by shikokuchuo** [reply]
Good observation. I think both the passive and active queues are going to be valuable. Passive can be used where this is an issue + it is simple and robust. Active can be used when it is merited.

**Comment by shikokuchuo** [comment, marked as answer]
This is now the definitive answer. Suggestion straight from Garrett D'Amore, author of NNG.

Instead of specifying a TCP URL, specify a Websocket URL starting `ws://`. As we can append arbitrary paths to ws addresses, suddenly we have no limits on the number of connections over one port!

As of [fce31f9](https://github.com/shikokuchuo/mirai/commit/fce31f9bb9aef7b235d82832913122c4ff934b02), I have implemented auto-incrementing the path numerically so we only need to supply one URL as before e.g.

```r
daemons("ws://:5555", nodes = 4)
```

Or specify a vector and use something more descriptive:

```r
daemons(c("ws://:5555/cpuserver", "ws://:5555/gpuserver"), nodes = 2)
```

Apparently what we were previously considering to solve this issue was, quote "similar to the old RPCBIND approach used 30 years ago"!!

**Reply by wlandau** [reply]
Wow! This is spectacular! Thank you so much for finding such a clean solution! I am so excited to move `crew` to websockets.

**Reply by wlandau** [reply]
I have a couple follow-up questions:

1. Would it still be possible to specify the client IP address, e.g. `daemons(c("ws://192.168.0.2:5555/cpuserver", "ws://192.168.0.2:5555/gpuserver"), nodes = 2)`?
2. Is it still possible for NNG to suggest a free port which I could then supply to the websockets I pass to `daemons()`? Totally fine if not, I can use `parallelly::freePort(ports = seq(from = 49152L, to = 65535L, by = 1L), default = NA_integer_)` (from https://github.com/HenrikBengtsson/parallelly/issues/101#issuecomment-1452716253).
3. I noticed you used "cpuserver" and "gpuserver" as suggested names. Does that mean you have plans to have the active queue send tasks to servers based on websocket descriptors? Again, totally fine if not, `crew` already has plans to handle heterogeneous workers using [`multicontrollers`](https://wlandau.github.io/crew/articles/multi_controller.html). (It does cost one active queue daemon process per worker group, but I am counting on the number of worker groups to be small.)

**Reply by shikokuchuo** [reply]
> I have a couple follow-up questions:
> 
>     1. Would it still be possible to specify the client IP address, e.g. `daemons(c("ws://192.168.0.2:5555/cpuserver", "ws://192.168.0.2:5555/gpuserver"), nodes = 2)`?
Yes supplying a vector of specific addresses should work.
>     2. Is it still possible for NNG to suggest a free port which I could then supply to the websockets I pass to `daemons()`? Totally fine if not, I can use `parallelly::freePort(ports = seq(from = 49152L, to = 65535L, by = 1L), default = NA_integer_)` (from [freePort() false negative? HenrikBengtsson/parallelly#101 (comment)](https://github.com/HenrikBengtsson/parallelly/issues/101#issuecomment-1452716253)).
Still works, but oddly the url is not updated. It is possible to get the port using the 'tcp-bound-port' option, but it makes it a bit unclean if I have to manually handle this to manipulate the URL when using ws. If freePort() works for you then I think that may be best for now.
> 
>     3. I noticed you used "cpuserver" and "gpuserver" as suggested names. Does that mean you have plans to have the active queue send tasks to servers based on websocket descriptors? Again, totally fine if not, `crew` already has plans to handle heterogeneous workers using [`multicontrollers`](https://wlandau.github.io/crew/articles/multi_controller.html). (It does cost one active queue daemon process per worker group, but I am counting on the number of worker groups to be small.)
Just for illustration I didn't really think about it. I am fine for Crew to handle this part. I think the logic will get complicated if I try this, whereas it is quite modular the way it is.

**Reply by wlandau** [reply]
> Still works, but oddly the url is not updated. It is possible to get the port using the 'tcp-bound-port' option, but it makes it a bit unclean if I have to manually handle this to manipulate the URL when using ws. If freePort() works for you then I think that may be best for now.

In that case I will happily use `parallelly::freePort()`.

> Just for illustration I didn't really think about it. I am fine for Crew to handle this part. I think the logic will get complicated if I try this, whereas it is quite modular the way it is.

Yes, please do not worry about this for `mirai`. I think the multicontroller solution in `crew` is clean and flexible, so I will continue to go with that.

**Reply by wlandau** [reply]
Just tested websockets on my end on different computers on the local network, and it works like a dream!

**Reply by shikokuchuo** [reply]
All thanks to gdamore! I use nanonext to consume websocket data all the time, but I didn't even know scalability protocols over ws was a thing!

**Reply by shikokuchuo** [reply]
> > Still works, but oddly the url is not updated. It is possible to get the port using the 'tcp-bound-port' option, but it makes it a bit unclean if I have to manually handle this to manipulate the URL when using ws. If freePort() works for you then I think that may be best for now.
> 
> In that case I will happily use `parallelly::freePort()`.
> 
I have actually added an extra check in [de40400](https://github.com/shikokuchuo/mirai/commit/de404008ba3a7cb7d67a75739fa2daa80d93436a) as having an inconsistent API between tcp and ws is rather unsatisfactory. Also it is pretty useless to have a zero port showing when the real port is something else. What this does is if the URL queried has a port of zero (should never be the case), it gets the actual port, disconnects and connects again immediately using this port explicitly.

So please feel free to use the NNG implementation instead of  `parallelly::freePort()`.

**Reply by wlandau** [reply]
Thanks! If I specify multiple ws:// websockets with 0 as the port, will mirai/NNG then assign all those websockets to the same actual port?

**Reply by shikokuchuo** [reply]
No just specify one URL in that case. It will append `/1`, `/2` etc. to the same random port.

If specifying a vector, it is difficult to reason whether you actually want the same port or not, so it is treated more literally.
